Add Export Dbus method.

xjuan opened the pull request:
This is a convenience method to export any object on the bus, this way after exporting it
the returneed DBus path can be used to create a proxy object ready to use on the client side.

Currently only properties are exposed but if needed its possible to export signals as well

https://phabricator.endlessm.com/T23926

tchx84 commented:
@xjuan the branch is crashing the toy-app here. Maybe I am doing something wrong? Here is what I am doing:

1. Run `com.endlessm.hackyballs` with this `clippy` branch.
2. I make sure the Export API is available with d-feet.
3. I run the following script:

```
const {Gio} = imports.gi;

const ViewName = 'view.JSContext.globalParameters';
const BusName = 'com.endlessm.hackyballs';
const AppObjectPath = '/com/endlessm/hackyballs';
const ClippyIface = `
<node xmlns:doc="http://www.freedesktop.org/dbus/1.0/doc.dtd">
  <interface name='com.endlessm.Clippy'>
    <method name='Export'>
      <arg type='s' name='object' />
      <arg type='s' name='path' direction='out'/>
    </method>
  </interface>
</node>
`;

const Proxy = Gio.DBusProxy.makeProxyWrapper(ClippyIface);
const proxy = new Proxy(Gio.DBus.session, BusName, AppObjectPath);
print(proxy.ExportSync('view.JSContext.globalParameters.backgroundImageIndex'));
```

The error I get on the app side:
```
(app.py:4): GLib-GIO-CRITICAL **: 08:53:12.456: g_dbus_object_skeleton_new: assertion 'g_variant_is_object_path (object_path)' failed
(app.py:4): GLib-GIO-CRITICAL **: 08:53:12.456: g_dbus_object_skeleton_add_interface: assertion 'G_IS_DBUS_OBJECT_SKELETON (object)' failed
/app/bin/com.endlessm.hackyballs: line 3:     4 Segmentation fault      (core dumped) python3 /app/share/toy/app.py com.endlessm.hackyballs
```

The error I get on the script side:
```
(gjs:27349): Gjs-WARNING **: JS ERROR: Gio.DBusError: GDBus.Error:org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
_proxyInvoker@resource:///org/gnome/gjs/modules/overrides/Gio.js:94:16
_makeProxyMethod/<@resource:///org/gnome/gjs/modules/overrides/Gio.js:124:16
@clippy.js:19:7

JS_EvaluateScript() failed
```

tchx84 commented:
UPDATE:

If I change the script above to use:
```
proxy.ExportSync('view') // just view
```
Then it doesn't crash, and I get an object path, but I can't see it with d-feet. And I see the following error in the app side:
```

(app.py:4): GLib-GIO-CRITICAL **: 09:13:08.998: g_dbus_object_manager_server_export_unlocked: assertion 'g_str_has_prefix (object_path, manager->priv->object_path_ending_in_slash)' failed

```

xjuan commented:
Right, first of all it should not crash, but in any case you used 'view.JSContext.globalParameters.backgroundImageIndex' which is not an object, in theory 'view.JSContext.globalParameters' should work

xjuan commented:
OK so I fixed one issue regarding object paths with JS objects names and also added a few error checks so that if you try to access a JS object property as if it where an object it will fail.

So now the proxy wrapper should work properly with JS objects!

joaquimrocha commented on src/clippy-dbus-wrapper.c:
```diff
@@ -0,0 +1,324 @@
+clippy_dbus_wrapper_skeleton_dbus_interface_get_properties (GDBusInterfaceSkeleton *skeleton)
+{
+  ClippyDbusWrapperPrivate *priv = CLIPPY_DBUS_WRAPPER_PRIVATE (skeleton);
+  g_autoptr (GVariantBuilder) b = NULL;
```
No space between the macro and the parenthesis needed.

tchx84 commented:
@xjuan yeah, is not crashing anymore, but still can't export the object properly. I am using this:

```
const {Gio, GLib} = imports.gi;

const ViewName = 'view.JSContext.globalParameters';
const BusName = 'com.endlessm.hackyballs';
const AppObjectPath = '/com/endlessm/hackyballs';
const ClippyIface = `
<node xmlns:doc="http://www.freedesktop.org/dbus/1.0/doc.dtd">
  <interface name='com.endlessm.Clippy'>
    <method name='Export'>
      <arg type='s' name='object' />
      <arg type='s' name='path' direction='out'/>
    </method>
  </interface>
</node>
`;

const Proxy = Gio.DBusProxy.makeProxyWrapper(ClippyIface);
const proxy = new Proxy(Gio.DBus.session, BusName, AppObjectPath);
print(proxy.ExportSync(ViewName));
```

And it prints `/com/endlessm/hackyballs/objects/view/JSContext/globalParameters`, but I can't see the object_path using d-feet.

On the app side, I see the following error `GLib-GIO-CRITICAL **: 09:22:16.345: g_dbus_object_manager_server_export_unlocked: assertion 'g_str_has_prefix (object_path, manager->priv->object_path_ending_in_slash)' failed
`

joaquimrocha commented on src/clippy.c:
```diff
@@ -480,43 +482,59 @@ clippy_emit (Clippy       *clip,
+               GError      **error)
+{
+  g_autoptr (GDBusObjectSkeleton) skeleton = NULL;
+  g_autoptr (ClippyDbusWrapper) interface = NULL;
```
The space between the macro and parenthesis again.

joaquimrocha commented on src/utils.c:
```diff
@@ -49,6 +49,55 @@ object_get_name (GObject *object)
+  g_autoptr(GVariant) variant = NULL;
+  g_autoptr(GArray) values = NULL;
+  GValue *instance_and_params;
+  guint i;
```
You should declare this in the loop declaration itself.

xjuan replied:
C89

joaquimrocha commented on src/utils.c:
```diff
@@ -235,6 +289,36 @@ app_get_object_info (GApplication *app,
+    return "v";
+  else if (type == G_TYPE_STRV)
+    return "as";
+  else if (type == G_TYPE_STRING || type == G_TYPE_GTYPE || type == G_TYPE_GSTRING ||
```
I wonder if these should be in a switch statement (and the rest in ifs), in case that helps readability.

xjuan replied:
well some types are not constants , so you cant really use a switch statement

tchx84 commented:
@xjuan it doesn't crash anymore and now I do see the object_path and properties from dfeet, but it looks like the values are not updating.. see example below:

After I export the object with the previous example (see previous replies). then I use this to play witht he properties:

```
from gi.repository import Gio, GLib

proxy = Gio.DBusProxy.new_sync(Gio.bus_get_sync(Gio.BusType.SESSION, None),
                                0,
                                None,
                                'com.endlessm.HackUnlock',
                                '/com/endlessm/HackUnlock/objects/view/JSContext/globalParameters',
                                'org.glib.GObject',
                                None)


def _print_mode():
    print(proxy.get_cached_property('mode'))
    return True

proxy.connect('g_properties_changed', _print_mode)
_print_mode()

GLib.MainLoop().run()
```

Then from the toy-app webkit console `globalParameters.mode = <to a new value>`, but the proxy doesn't emit the signal, nor the value is updated.

joaquimrocha commented:
@tchx84 It seems like you're not using the right interface in the DBus proxy call. I have changed your script to the following and I am able to get the properties changed signal:
```
from gi.repository import Gio, GLib

clippy = Gio.DBusProxy.new_sync(Gio.bus_get_sync(Gio.BusType.SESSION, None),
                                0,
                                None,
                                'com.endlessm.HackUnlock',
                                '/com/endlessm/HackUnlock',
                                'com.endlessm.Clippy',
                                None)

def _print_mode(obj, props_changed, invalidated):
    print(props_changed.unpack())
    return True

clippy.Export('(s)', 'view.JSContext.globalParameters')

proxy = Gio.DBusProxy.new_sync(Gio.bus_get_sync(Gio.BusType.SESSION, None),
                                0,
                                None,
                                'com.endlessm.HackUnlock',
                                '/com/endlessm/HackUnlock/objects/view/JSContext/globalParameters',
                                'com.endlessm.Clippy.Object',
                                None)

print(proxy.get_cached_property_names())
print(proxy.get_cached_property('mode'))

proxy.connect('g_properties_changed', _print_mode)

GLib.MainLoop().run()
```

tchx84 commented:
@joaquimrocha thanks, we changed the PR yesterday with @xjuan , these scripts were out-of-date. The issue seemed to be that clippy was not emitting the right signal when the properties changed (and we also changed the iface name).

tchx84 commented on src/clippy-dbus-wrapper.c:
```diff
@@ -0,0 +1,397 @@
+  g_variant_builder_add (&builder, "{sv}", pspec->name, variant_new_value (&value));
+  g_variant_builder_add (&invalidated_builder, "s", pspec->name);
+
+  g_message ("%s %s", __func__, pspec->name);
```
forgot to remove?
